### PR TITLE
Add --default-time-limit

### PR DESCRIPTION
**Pull request by epdenouden**
Implementation for #2085 

This is a convenience feature for people managing large(r) and potentially slow legacy test collections. Classic example: tests for database integration that have huge `@dataprovider` sets or when using PHPUnit to test webservices/APIs.

Note: I tried running around with my laptop, however I did not manage to validate the timer at relativistic speeds.

### CODE REVIEW

**Comment by epdenouden** [comment]
Ah yes, I was afraid it would not work on Windows yet. Will fix that later today.

**Comment by epdenouden** [comment]
@sebastianbergmann I'd like to fix covecov before merging, too

File: src/TextUI/Command.php [hunk 1]
```diff
@@ -1111,6 +1117,7 @@ protected function showHelp(): void
   --disallow-test-output      Be strict about output during tests
   --disallow-resource-usage   Be strict about resource usage during small tests
   --enforce-time-limit        Enforce time limit based on test size
+  --default-time-limit=<sec>  Maximum running time in seconds for unsized tests
```
Comment (reinholdfuereder) [inline src/TextUI/Command.php]: I am not perfectly happy with that help text: (a) I personally think something like "Timeout in seconds ..." is more common? (b) Is "unsized tests" also understandable for others? How short/compact must the help text be: is "... for tests without either `@small`, `@medium` or `@large`" too long?
(Sorry, if this is too picky)

Comment (epdenouden) [reply src/TextUI/Command.php]: First of all: feel free to be as picky as you want, it was your idea after all!  There is no deadline pressure from management, so we can take all the time we need to get it from 'good enough' to 'exactly what I needed'. :-)

I agree with your criticism of the help text. It isn't great as I tried to get all the relevant information in less than 80 characters. Mentioning the three sizes of tests might not fit. I do agree that using _timeout_ would be nice. The config flag is already called `*-time-limit` and adding `timeout` will make it extra clear what it does. I'll commit an improved help text, let me know what you think.

**Comment by epdenouden** [comment]
@sebastianbergmann Getting the diff coverage above the threshold would require `php-invoker` or adding a fake `Invoker`. For such a small feature in the `TestRunner`, are you okay with it as-is?

File: src/TextUI/Command.php (hunk 1 above)
Comment (reinholdfuereder) [reply src/TextUI/Command.php]: Thanks, I like this!

**Comment by karmakoder** [comment]
@epdenouden @sebastianbergmann @reinholdfuereder any update here? This change makes a lot of sense for our legacy phpunit suites and a merge will be much appreciated.

**Comment by epdenouden** [comment]
@karmakoder No news from my side. I agree it would be nice to see it merged.

**Comment by bc-garybottom** [comment]
@sebastianbergmann huge +1 here as well, working with our php code base for functional and unit tests, particularly for UI issues having this functionality would be huge.  Running in remote CI/CD SaaS solutions we have the need to gracefully timeout more and more to achieve automation autonomy without chasing down red herrings in our tests or CI/CD tool.  We're currently using phpunit 6.0.13 for our test suite, for reference, which works great except having the ability to skip tests that are idle to rerun them later with this functionality.

**Comment by epdenouden** [comment]
@bc-garybottom This pull request is for version 7. If it gets merged I'd be happy to backport it to 6.x for you if @sebastianbergmann would accept a PR.

**Comment by sebastianbergmann** [comment]
Sorry, but no new features for PHPUnit 6.

**Comment by karmakoder** [comment]
NM. Figured what was wrong with my local setup. All good.

**Comment by epdenouden** [comment]
@karmakoder No problem. Let me know if you run into bugs or edge cases I didn't think of yet.

**Comment by reinholdfuereder** [comment]
@karmakoder Ad "Any update here?": Not really

- there is still my small pending review request (also mentioned/outlined in https://github.com/sebastianbergmann/phpunit/issues/2085#issuecomment-409466372)
- and actually based on your experience mentioned in https://github.com/sebastianbergmann/phpunit/pull/3224#issuecomment-418922883 one suggestion: what about replacing the `writeMessage` with a more defensive failing of the test?

**Comment by epdenouden** [comment]
@reinholdfuereder Apologies for missing your recent comment, missed that completely! Let me give that a think.

**Comment by epdenouden** [comment]
@sebastianbergmann if you have time to look at the PR during the Code Sprint, that'd be lovely 🔬

File: src/Framework/TestResult.php [hunk 2]
```diff
@@ -139,6 +139,11 @@ class TestResult implements Countable
      */
     protected $beStrictAboutResourceUsageDuringSmallTests = false;
 
+    /**
+     * @var int
+     */
+    protected $defaultTimeLimit = 0;
```
Comment (sebastianbergmann) [inline src/Framework/TestResult.php]: New attributes should be `private`.

File: src/TextUI/Command.php [hunk 3]
```diff
@@ -84,6 +84,7 @@ class Command
         'disallow-test-output'      => null,
         'disallow-resource-usage'   => null,
         'disallow-todo-tests'       => null,
+        'default-time-limit='       => null,
```
Comment (sebastianbergmann) [inline src/TextUI/Command.php]: For consistency, I would rather not add an option like `--default-time-limit` to set the default time limit on the CLI. The timeouts for `@small`, `@medium`, and `@large` can also only be configured in the configuration file.

File: src/Framework/TestResult.php (hunk 2 above)
Comment (epdenouden) [reply src/Framework/TestResult.php]: Fixing

File: src/TextUI/Command.php (hunk 3 above)
Comment (epdenouden) [reply src/TextUI/Command.php]: Yes, that is true. I asked @reinholdfuereder what the exact use case would be and this time limit is meant to be for tests that are _not marked_ with a `@size`: https://github.com/sebastianbergmann/phpunit/issues/2085#issuecomment-405539959

I know a bunch of older test collections with legacy would benefit from this.

Comment (localheinz) [reply src/TextUI/Command.php]: Not sure about the trailing`=`, what do you think?

Comment (epdenouden) [reply src/TextUI/Command.php]: The `=` is required for `--key=value`, right? I have not found a clean way of doing `--key[=value]`.

Comment (sebastianbergmann) [reply src/TextUI/Command.php]: The `=` means that the option requires a value.

Comment (localheinz) [reply src/TextUI/Command.php]: Thank you for clarifying, @sebastianbergmann!

**Comment by karmakoder** [comment]
@epdenouden After doing several rounds of testing, am learning that this default-time-limit doesnt always take effect for our tests. I basically backported your changes in to phpunit-6.0.13 but the behavior of tests getting marked as failed doesnt seem to be working 100% of the time. Any pointers on what could be the reason? Unfortunately, we cannot upgrade to latest phpunit version as several of our dependencies don't yet support higher versions of php and phpunit. If it helps, we can connect offline to discuss this. Thx in advance.

**Pull request merged**